internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,175 @@
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	if err == nil {
+		fmt.Printf("%T \n %v", environmentIDInt, environmentIDInt)
+	}
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return diag.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	name := d.Get("name").(string)
+	options := gitlab.EditEnvironmentOptions{
+		Name: &name,
+	}
+	if d.HasChange("external_url") {
+		options.ExternalURL = gitlab.String(d.Get("external_url").(string))
+	}
+
+	log.Printf("[DEBUG] Project %s update gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	if _, _, err := client.Environments.EditEnvironment(project, environmentIDInt, &options); err != nil {
+		return diag.Errorf("error editing gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return diag.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	log.Printf("[DEBUG] Project %s delete gitlab project-level environment %v", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	_, err = client.Environments.StopEnvironment(project, environmentIDInt)
```

Comment:
```suggestion
	_, err = client.Environments.StopEnvironment(project, environmentIDInt, gitlab.WithContext(ctx))
```

Reply:
Is my assumption correct, that an environment isn't stopped implicitly if `DELETE` is called?

Would it make sense to add an additional attribute to control if the environment should be stopped prior deletion as like a safety measure to not delete environment which haven't been stopped ... ?

At least we should document that behavior in the appropriate box, maybe something like:

```markdown
~> **Behavior** The environment is stopped before deletion
```

Reply:
I believe you are correct based off of the documentation [link](https://docs.gitlab.com/ee/ci/environments/#stop-an-environment). I think that might be too much considering that a stop job would have to be implemented for that to work correctly on top of sometimes if an environment in GitLab is deleted whether it be for renaming or something else, sometimes people might not want to stop the underlying infrastructure itself.  For example if someone wants to rename an environment it has to be deleted and recreated [link](https://docs.gitlab.com/ee/ci/environments/#rename-an-environment)

Reply:
If you think it might be worthwhile we could make a flag that is false by default and then people have to toggle it to true to enforce the environment is stopped before the resource can be deleted.  I wouldn't have it toggled on by default though